### PR TITLE
push \xff correctly

### PR DESCRIPTION
**Pull request by Hecate2**
Close https://github.com/neo-project/neo-devpack-dotnet/issues/794
Close https://github.com/neo-project/neo-devpack-dotnet/issues/856

### CODE REVIEW

**Comment by Hecate2** [comment]
[0x00-0xff.txt](https://github.com/user-attachments/files/17595392/0x00-0xff.txt)
A text file for all chars from 0x00 to 0xff, split by '\n'

**Comment by shargon** [comment]
> [0x00-0xff.txt](https://github.com/user-attachments/files/17595392/0x00-0xff.txt) A text file for all chars from 0x00 to 0xff, split by '\n'

Add it as ut?

**Comment by Hecate2** [comment]
> > [0x00-0xff.txt](https://github.com/user-attachments/files/17595392/0x00-0xff.txt) A text file for all chars from 0x00 to 0xff, split by '\n'
> 
> Add it as ut?

Done in https://github.com/neo-project/neo-devpack-dotnet/pull/1219/commits/ba53357262da864ca9a9cc7119197d06933889f3
```python
for i in range(256):
    print(f"\\x{hex(i)[2:].zfill(2)}", end='')
```

**Review by shargon — COMMENTED** [review]
Test don't pass @Hecate2

**Comment by Hecate2** [comment]
> Test don't pass @Hecate2

Encoding.ASCII.GetBytes does not work for bytes >= 128. Those bytes are non-ascii.

File: tests/Neo.Compiler.CSharp.UnitTests/UnitTest_ByteString.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+        public static string LiteralWithOtherChar() => "你好\x00\x80\xff";
+
+        [TestMethod]
+        public void Test_Literal00ToFF()
+        {
+            CollectionAssert.AreEqual(Contract.Literal00ToFF(), Encoding.ASCII.GetBytes(Literal00ToFF()));
```
Comment (shargon) [inline tests/Neo.Compiler.CSharp.UnitTests/UnitTest_ByteString.cs:17]:
```suggestion
            CollectionAssert.AreEqual(Contract.Literal00ToFF(), Encoding.UTF8.GetBytes(Literal00ToFF()));
```

File: src/Neo.Compiler.CSharp/MethodConvert/Helpers/StackHelpers.cs [hunk 2]
```diff
@@ -114,9 +115,29 @@ private Instruction Push(BigInteger number)
         return instruction;
     }
 
+    /// <summary>
+    /// If all char in the input is no greater than byte.MaxValue == 255, get the byte for each char
+    /// If any char in the input is greater than byte.MaxValue == 255, encode in UTF8
+    /// This ensures "a\xff\x80\x79\x00" accurately translated to 0x61 0xff 0x80 0x79 0x00
+    /// </summary>
+    /// <param name="s"></param>
+    /// <returns></returns>
```
Comment (shargon) [inline src/Neo.Compiler.CSharp/MethodConvert/Helpers/StackHelpers.cs:124]:
```suggestion
    /// <param name="s">String value to push</param>
    /// <returns>Instruction</returns>
```

**Comment by Hecate2** [comment]
```csharp
        [TestMethod]
        public void Test_Literal00ToFF()
        {
            byte[] result = Contract.Literal00ToFF()!;
            Assert.AreEqual(result.Length, 256);
            for (int i = 0; i <= 255; ++i)
                Assert.AreEqual(i, result[i]);
        }
```

**Comment by shargon** [comment]
> ```cs
>         [TestMethod]
>         public void Test_Literal00ToFF()
>         {
>             byte[] result = Contract.Literal00ToFF()!;
>             Assert.AreEqual(result.Length, 256);
>             for (int i = 0; i <= 255; ++i)
>                 Assert.AreEqual(i, result[i]);
>         }
> ```

Could you change it?

**Review by shargon — APPROVED** [review]

**Pull request merged**